modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/near/IgniteTxExceptionNodeFailTest.java

```diff
@@ -0,0 +1,207 @@
+import org.apache.ignite.internal.TestRecordingCommunicationSpi;
+import org.apache.ignite.internal.managers.communication.GridIoMessage;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteInClosure;
+import org.apache.ignite.plugin.extensions.communication.Message;
+import org.apache.ignite.spi.IgniteSpiException;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.transactions.Transaction;
+import org.apache.ignite.transactions.TransactionHeuristicException;
+import org.junit.Test;
+
+import static org.apache.ignite.cache.CacheWriteSynchronizationMode.FULL_SYNC;
+import static org.apache.ignite.cache.CacheWriteSynchronizationMode.PRIMARY_SYNC;
+
+/**
+ * Tests check a result of commit when a node fail before
+ * send {@link GridNearTxFinishResponse} to transaction coodinator
+ */
+public class IgniteTxExceptionNodeFailTest extends GridCommonAbstractTest {
+    /** Spi for node0 */
+    private SpecialSpi spi0;
+
+    /** Spi for node1 */
+    private SpecialSpi spi1;
+
+    /** syncMode */
+    private static CacheWriteSynchronizationMode syncMode;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        DataStorageConfiguration dsConfig = new DataStorageConfiguration()
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                .setPersistenceEnabled(true));
+
+        SpecialSpi spi = new SpecialSpi();
+
+        cfg.setCommunicationSpi(spi);
+
+        if (igniteInstanceName.contains("0"))
+            spi0 = spi;
+
+        if (igniteInstanceName.contains("1"))
+            spi1 = spi;
+
+        return cfg
+            .setDataStorageConfiguration(dsConfig)
+            .setCacheConfiguration(new CacheConfiguration("cache")
+                .setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+                .setWriteSynchronizationMode(syncMode).setBackups(0));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        FileUtils.deleteDirectory(new File(U.defaultWorkDirectory()));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+    }
+
+    /**
+     * Test with {@link CacheWriteSynchronizationMode#PRIMARY_SYNC}
+     * @throws Exception
+     */
+    @Test
+    public void testNodeFailWithPrimarySync() throws Exception {
+        testNodeFail(PRIMARY_SYNC);
+    }
+
+    /**
+     * Test with {@link CacheWriteSynchronizationMode#FULL_SYNC}
+     * @throws Exception
+     */
+    @Test
+    public void testNodeFailWithFullSync() throws Exception {
+        testNodeFail(FULL_SYNC);
```

Comment:
What's the point in testing full sync mode with zero backups ?

Reply:
Because it is possible to set FULL_SYNC with zero backups. The fix has code for this case.